base/linalg/diagonal.jl

```diff
@@ -350,8 +350,8 @@ function pinv{T}(D::Diagonal{T}, tol::Real)
 end
 
 #Eigensystem
-eigvals(D::Diagonal{<:Number}) = D.diag
-eigvals(D::Diagonal) = [eigvals(x) for x in D.diag] #For block matrices, etc.
+eigvals(D::Diagonal{<:Number}) = sort(D.diag, by=eigsortby)
+eigvals(D::Diagonal) = sort!(vcat((eigvals(x) for x in D.diag)...),by=eigsortby) #For block matrices, etc.
```

Comment:
That would be worse, I think, because it would call `vcat` pairwise over and over and generate lots of temporary arrays.